validator/db/kv/attestation_history_new.go

```diff
@@ -0,0 +1,146 @@
+package kv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+const (
+	// The size of each data entry in bytes for the source epoch (8 bytes) and signing root (32 bytes).
+	uint64Size             = 8
+	latestEpochWrittenSize = uint64Size
+	targetSize             = uint64Size
+	sourceSize             = uint64Size
+	signingRootSize        = 32
+	historySize            = targetSize + sourceSize + signingRootSize
+	minimalSize            = latestEpochWrittenSize
+)
+
+// AttestationHistoryNew stores the historical attestation data needed
+// for protection of validators.
+type AttestationHistoryNew struct {
+	TargetToSource     map[uint64]*HistoryData
+	LatestEpochWritten uint64
+}
+
+// HistoryData stores the needed data to confirm if an attestation is slashable
+// or repeated.
+type HistoryData struct {
+	Source      uint64
+	SigningRoot []byte
+}
+
+func newAttestationHistoryArray(target uint64) []byte {
+	enc := make([]byte, latestEpochWrittenSize+(target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize+historySize)
+	return enc
+}
+
+func sizeChecks(enc []byte) error {
+	if enc == nil || len(enc) < minimalSize {
+		return fmt.Errorf("encapsulated data size: %d is smaller then minimal size: %d", len(enc), minimalSize)
+	}
+	if (len(enc)-minimalSize)%historySize != 0 {
+		return fmt.Errorf("encapsulated data size: %d is not a multiple of entry size: %d", len(enc), historySize)
+	}
+
+	return nil
+}
+
+func getLatestEpochWritten(ctx context.Context, enc []byte) (uint64, error) {
+	if err := sizeChecks(enc); err != nil {
+		return 0, err
+	}
+	return bytesutil.FromBytes8(enc[:latestEpochWrittenSize]), nil
+}
+
+func setLatestEpochWritten(ctx context.Context, enc []byte, latestEpochWritten uint64) ([]byte, error) {
+	if err := sizeChecks(enc); err != nil {
+		return nil, err
+	}
+	copy(enc[:latestEpochWrittenSize], bytesutil.Uint64ToBytesLittleEndian(latestEpochWritten))
+	return enc, nil
+}
+
+func getTargetData(ctx context.Context, enc []byte, target uint64) (*HistoryData, error) {
+	if err := sizeChecks(enc); err != nil {
+		return nil, err
+	}
+	cursor := (target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize + latestEpochWrittenSize
```

Comment:
W/o comment, calculation above is kind of cryptic. Even one elaborating line can help a lot here. Although, it might be well related to my limited understanding of this part of the code.

Reply:
thanks